File: ps3/Template_Ex_6/Template_Ex_6/Functions/Fsys_Reiter.jl
include("EGM.jl")
include("Young.jl")

"""
    Fsys_Reiter(Xhat, XhatPrime, XSS, mpar, par, idSS,idDev,gri,meshes,Π)

Compute deviations from equilibrium conditions

    * `Xhat`,`XhatPrime`: deviations of present and future variables from steady state
    * `XSS`: steady state values (in logs)
    * `mpar`,`par`: parameter structures
    * `idSS`,`idDev`: index structures for steady state and deviation values
    * `gri`,`meshes`: grids and meshes for income (z) and capital (k)
    * `Π`: transition probability matrix of income (z)

Returns

    * `F`: vector of deviations from equilibrium conditions
"""
function Fsys_Reiter(Xhat, XhatPrime, XSS, mpar, par, idSS,idDev,gri,meshes,Π)
    # expand X around steady state
    
    X      = exp.(Xhat + XSS)
    XPrime = exp.(XhatPrime+ XSS)
    
    Dist            = zeros(eltype(Xhat),mpar.nk, mpar.nz)
    Dist[1:end-1]   = Xhat[idDev.Dist] + XSS[idSS.Dist]
    Dist[end]       = 1 - sum(Dist[:])
    Dist            = Dist[:]
    
    DistPrime            = zeros(eltype(Xhat),mpar.nk, mpar.nz)
    DistPrime[1:end-1]   = XhatPrime[idDev.Dist] + XSS[idSS.Dist]
    DistPrime[end]       = 1 - sum(DistPrime[:])
    DistPrime            = DistPrime[:]
    #--------------------------------------------------------------------------------------------------------------
    # Aggregate Model
    #--------------------------------------------------------------------------------------------------------------
    F = zeros(eltype(Xhat),length(Xhat))
    # Capital summary
    F[idDev.K] = log(X[idSS.K]) - log(dot(xxx,xxx))   
    # Output
    F[idDev.Y] = log(X[idSS.Y]) - log(X[idSS.Z]*par.N^(1-par.α)*X[idSS.K]^par.α) 
    # TFP
    F[idDev.Z] = log(XPrime[idSS.Z]) - par.ρ_Z*log(xxx)
    # Interest rates
    F[idDev.R] = log(X[idSS.R]) - log(xxx)
    # Consumption
    F[idDev.C] = log(X[idSS.C]) - log(xxx) 
    # Investment
    F[idDev.I] = log(X[idSS.I]) - log(xxx)
    
    #--------------------------------------------------------------------------------------------------------------
    # Individual Consumption Policies
    #--------------------------------------------------------------------------------------------------------------
    mutil(c)    = 1.0 ./c  # Marginal utility
    invmutil(mu) = 1.0 ./mu # inverse marginal utility
    if par.γ != 1.0
        mutil(c)     = 1.0 ./(c .^par.γ) # Marginal utility
        invmutil(mu) = 1.0 ./(mu .^(1.0 ./par.γ)) # inverse marginal utility
    end
    w        = X[idSS.Z]*par.N^(-par.α)*X[idSS.K]^par.α*(1-par.α)
    @set! meshes.z = meshes.z*w
    cpol,kprime = EGM(XPrime[idSS.cpol],mutil,invmutil,X[idSS.R],XPrime[idSS.R],par,mpar,Π,meshes,gri)
    
    F[idDev.cpol]= log.(xxx) - log.(xxx)
    
    #--------------------------------------------------------------------------------------------------------------
    # Individual Consumption Policies
    #--------------------------------------------------------------------------------------------------------------
    Γ = YoungSmall(kprime,gri,mpar,Π)
    
    DistNewPrime = xxx
    
    F[idDev.Dist] = DistPrime[1:end-1] - DistNewPrime[1:end-1]
    
    return F
end

File: ps3/Ex6/Ex6/Functions/Fsys_Reiter.m
function F = Fsys_Reiter(Xhat, XhatPrime, XSS, mpar, par, id,gri,meshes,P)
% expand X around steady state

X      = exp(Xhat + XSS);
XPrime = exp(XhatPrime+ XSS);

Dist            = zeros(mpar.nk, mpar.nz);
Dist(1:end-1)   = Xhat(id.Dev.Dist) + XSS(id.SS.Dist);
Dist(end)       = 1 - sum(Dist(:));
Dist            = Dist(:);

DistPrime            = zeros(mpar.nk, mpar.nz);
DistPrime(1:end-1)   = XhatPrime(id.Dev.Dist) + XSS(id.SS.Dist);
DistPrime(end)       = xxx;
DistPrime            = DistPrime(:);
%--------------------------------------------------------------------------------------------------------------
% Aggregate Model
%--------------------------------------------------------------------------------------------------------------

% Capital summary
F(id.Dev.K) = log(X(id.SS.K)) - log(dot(xxx,xxx));   
% Output
F(id.Dev.Y) = log(X(id.SS.Y)) - log(X(id.SS.Z)*par.N^(1-par.alpha)*X(id.SS.K)^par.alpha); 
% TFP
F(id.Dev.Z) = log(XPrime(id.SS.Z)) - par.rho_Z*log(xxx);
% Interest rates
F(id.Dev.R) = log(X(id.SS.R)) - log(xxx);
% Consumption
F(id.Dev.C) = log(X(id.SS.C)) - log(xxx); 
% Investment
F(id.Dev.I) = log(X(id.SS.I)) - log(xxx); 

%--------------------------------------------------------------------------------------------------------------
% Individual Consumption Policies
%--------------------------------------------------------------------------------------------------------------
if par.gamma ==1
    mutil    = @(c) 1./c;  % Marginal utility
    invmutil = @(mu) 1./mu;% inverse marginal utility
else
    mutil    = @(c) 1./(c.^par.gamma); % Marginal utility
    invmutil = @(mu) 1./(mu.^(1./par.gamma)); % inverse marginal utility
end
w        = X(id.SS.Z)*par.N^(-par.alpha)*X(id.SS.K)^par.alpha*(1-par.alpha);
meshes.z = meshes.z*w;
[cpol,kprime] = EGM(xxx,X(id.SS.R),XPrime(id.SS.R),par,mpar,P,meshes,gri, mutil, invmutil);

F(id.Dev.cpol)= log(xxx) - log(xxx); 

%--------------------------------------------------------------------------------------------------------------
% Individual Consumption Policies
%--------------------------------------------------------------------------------------------------------------
Gamma = Young(kprime,gri,mpar,P);

DistNewPrime = xxx;

F(id.Dev.Dist) = DistPrime(1:end-1) - DistNewPrime(1:end-1);

end